R/submission_tmpl.R

```diff
@@ -265,3 +300,57 @@ subset_complete_cases <- function(tmpl_df) {
   )
   tmpl_df[compl_cases, ]
 }
+
+# This function handles issuing deprecation warnings for older arguments
+# and returns a config_tasks list according to the input argument.
+switch_get_config <- function(hub_con, config_tasks, path) {
+  input_arg <- rlang::check_exclusive(hub_con, config_tasks, path,
+    .frame = parent.frame()
+  )
+  switch(input_arg,
+    hub_con = {
+      # Signal the deprecation to the user
+      lifecycle::deprecate_warn(
+        "0.11.0",
+        "hubValidations::submission_tmpl(hub_con = )",
+        "hubValidations::submission_tmpl(hub_path = )"
+      )
+      checkmate::assert_class(hub_con, classes = "hub_connection")
+      attr(hub_con, "config_tasks")
+    },
+    config_tasks = {
+      lifecycle::deprecate_warn(
+        "0.11.0",
+        "hubValidations::submission_tmpl(config_tasks = )",
+        "hubValidations::submission_tmpl(hub_path = )"
+      )
+      checkmate::assert_list(config_tasks)
+    },
+    path = {
+      # Read config from cloud hub
+      if (inherits(path, "SubTreeFileSystem")) {
+        if (hubUtils::is_s3_base_fs(path)) {
+          return(read_config(path))
+        }
+        return(read_config_file(path))
+      }
+      # Read config file from URL
+      if (hubUtils::is_url(path)) {
+        if (hubUtils::is_github_repo_url(path)) {
+          return(read_config(path))
+        }
+        return(read_config_file(path))
+      }
+      # Read local config
+      if (!fs::file_exists(path)) {
+        cli::cli_abort("{.arg path} {.file {path}} does not exist.",
+          call = rlang::caller_env(1)
+        )
+      }
+      if (fs::is_dir(path)) {
+        return(read_config(path))
+      }
+      read_config_file(path)
```

Comment:
Since much of the heavy lifting is done inside `read_config()` and `read_config_file()`, the main task is to determine if we have a directory or a file. This can be achieved with two indicator variables. 


```suggestion
      is_s3_dir <- inherits(path, "SubTreeFileSystem") && hubUtils::is_s3_base_fs(path)
      is_local_dir <- is.character(path) && fs::path_ext(path) == ""
      if (is_s3_dir || is_local_dir) {
        read_config(path)
      } else {
        read_config_file(path)
      }
```

Reply:
‼ This will not pass tests until https://github.com/hubverse-org/hubUtils/pull/211 is merged

Nice simplification! There however exceptions that this wasn't handling well. For example, if someone erroneously supplied `"https://github.com/hubverse-org/schemas/tree/main/v5.0.0"` as a path, it was identified as path and `read_config_file()` was invoked. That's why I added an additional check which catches that and returns a more informative error message.

`is.character(path) && fs::path_ext(path) == ""` is still a bit hacky as if a URL like `"https://raw.githubusercontent.com/hubverse-org/example-simple-forecast-hub/refs/heads/main/README"` is supplied, it would trigger `read_config()` erroneously. Unfortunately the only method I found for ensuring the correct read function is deployed requires checking the `Content-Type` to see if it’s a file. I've refrained from adding such a function but have added a test for that exception that would be nice to handle better at some point (now it just fails because it interprets the URL as a folder and creates a path to tasks.json that does not exists).

Reply:
> Nice simplification! There are however exceptions that this wasn't handling well. For example, if someone erroneously supplied `"https://github.com/hubverse-org/schemas/tree/main/v5.0.0"` as a path, it was identified as a file and `read_config_file()` was invoked. That's why I added an additional check which catches that and returns a more informative error message.
>
> `is.character(path) && fs::path_ext(path) == ""` is still a bit hacky as if a URL like `"https://raw.githubusercontent.com/hubverse-org/example-simple-forecast-hub/refs/heads/main/README"` or `"https://raw.githubusercontent.com/hubverse-org/example-simple-forecast-hub/refs/heads/main/LICENSE"` is supplied, it would trigger `read_config()` erroneously.

These are fair points and I definitely agree that the directory check is hacky, but I take solace in the hope that these situations are vanishingly rare and are easily fixed.